src/python/pants/backend/python/dependency_inference/rules_test.py

```diff
@@ -109,3 +116,60 @@ def run_dep_inference(address: Address) -> InferredDependencies:
         assert run_dep_inference(generated_subtarget_address) == InferredDependencies(
             [Address("src/python", target_name="app.py", generated_base_target_name="python")]
         )
+
+    def test_infer_python_inits(self) -> None:
+        options_bootstrapper = create_options_bootstrapper(
+            args=["--backend-packages=pants.backend.python", "--source-root-patterns=src/python",]
+        )
+
+        self.create_file("src/python/root/__init__.py")
+        self.add_to_build_file("src/python/root", "python_library()")
+
+        self.create_file("src/python/root/mid/__init__.py")
+        self.add_to_build_file("src/python/root/mid", "python_library()")
+
+        self.create_file("src/python/root/mid/leaf/__init__.py")
+        self.add_to_build_file("src/python/root/mid/leaf", "python_library()")
+
+        def run_dep_inference(address: Address) -> InferredDependencies:
+            target = self.request_single_product(WrappedTarget, address).target
+            return self.request_single_product(
+                InferredDependencies,
+                Params(InferInitDependencies(target[PythonSources]), options_bootstrapper),
+            )
+
+        assert run_dep_inference(Address.parse("src/python/root/mid/leaf")) == InferredDependencies(
+            [
+                Address("src/python/root", "__init__.py", generated_base_target_name="root"),
+                Address("src/python/root/mid", "__init__.py", generated_base_target_name="mid"),
+            ]
+        )
+
+    def test_infer_python_conftests(self) -> None:
+        options_bootstrapper = create_options_bootstrapper(
+            args=["--backend-packages=pants.backend.python", "--source-root-patterns=src/python",]
+        )
+
+        self.create_file("src/python/root/conftest.py")
+        self.add_to_build_file("src/python/root", "python_library(sources=['conftest.py'])")
+
+        self.create_file("src/python/root/mid/conftest.py")
+        self.add_to_build_file("src/python/root/mid", "python_library(sources=['conftest.py'])")
+
+        self.create_file("src/python/root/mid/leaf/conftest.py")
+        self.create_file("src/python/root/mid/leaf/this_is_a_test.py")
```

Comment:
No need to change this, only an FYI. You can use `self.create_files` when they have the same parent directory.